File: tools/generate_filelist/generate_filelist.py
#!/usr/bin/python

import sys, os, getopt
import random

def usage():
    print "Usage: generate_filelist -d rootdir -f filter -o filelist [-n sample_num]"
    print "  rootdir:\t file root directory"
    print "  filter:\t file name filter, eg: \"*.png\" "
    print "  filelist:\t output file name"
    print "  sample_num:\t sample num (>0)"

def generate_filelist(argv):
    if len(argv) != 6 and len(argv) != 8:
        usage()
        sys.exit(1)

    rootdir = ''
    filter = ''
    filelist = ''
    sample_num = 0
    try:
        opts, args = getopt.getopt(argv, "d:f:o:n:", ["rootdir=", "filter=", "filelist=", "sample_num="])
    except getopt.GetoptError:
        usage()
        sys.exit(1)

    for opt, arg in opts:
        if opt in ("-d", "--rootdir"):
            rootdir = arg
        elif opt in ("-f", "--filter"):
            filter = arg
        elif opt in ("-o", "--output"):
            filelist = arg
        elif opt in ("-n", "--sample_num"):
            sample_num = int(arg)
            if sample_num < 0:
                usage()
                sys.exit(1)
        else:
            usage()
            sys.exit(1)

    lists = []
    os.chdir(rootdir)

    for root, dirs, files in os.walk('.'):
        for file in files:
            if file.endswith(filter[1:]):
                lists.append(os.path.join(root, file))

    list = range(0, len(lists))
    indexes = []
    if sample_num == 0:
        indexes = list
    else:
        indexes = random.sample(list, sample_num)

    with open(filelist, 'w') as f:
        for index in indexes:
            f.write(lists[index][2:] + '\n')


if __name__ == "__main__":
    generate_filelist(sys.argv[1:])

File: tools/generate_filelist.rb
#!/usr/bin/ruby

require 'optparse'

argv_size = ARGV.size
options = {}

optparse = OptionParser.new do |opts|
    opts.banner = "Usage: generate_filelist -d rootdir -f filter -o filelist [-n sample_num]"
    opts.on('-d', '--rootdir rootdir', 'file root directory') { |v| options[:rootdir] = v }
    opts.on('-f', '--filter filter', 'file name filter') { |v| options[:filter] = v }
    opts.on('-o', '--filelist filelist', 'output file name') { |v| options[:filelist] = v }
    opts.on('-n', '--sample_num sample_num', 'sample num') { |v| options[:sample_num] = v }
end

begin
    optparse.parse!
rescue OptionParser::InvalidArgument, OptionParser::InvalidOption, OptionParser::MissingArgument
    puts $!.to_s
    puts optparse
    exit
end

if argv_size != 6 && argv_size != 8
    puts optparse
end

Dir.chdir(options[:rootdir])
lists =  Dir.glob("**/"+options[:filter])

file = File.new(options[:filelist], "w")
lists.each do |image|
    file.puts image
end

file.close

